File: week6/cpp/decorator_pattern/Coffee.cpp
#include "Coffee.hpp"

// 기본 커피의 가격을 반환
int Coffee::cost() const {
    return 5;
}

// 커피 데코레이터 생성자
CoffeeDecorator::CoffeeDecorator(Coffee* coffee) : coffee(coffee) {}

// 커피 데코레이터의 cost 메서드
int CoffeeDecorator::cost() const {
    return coffee->cost();
}

// 우유 데코레이터의 cost 메서드
MilkDecorator::MilkDecorator(Coffee* coffee) : CoffeeDecorator(coffee) {}

int MilkDecorator::cost() const {
    return coffee->cost() + 2;  // 우유는 2가 추가됨
}

// 설탕 데코레이터의 cost 메서드
SugarDecorator::SugarDecorator(Coffee* coffee) : CoffeeDecorator(coffee) {}

int SugarDecorator::cost() const {
    return coffee->cost() + 1;  // 설탕은 1이 추가됨
}

File: week6/kotlin/decorator_pattern/Coffee.kt
// 기본 커피 클래스
open class Coffee {
    open fun cost(): Int {
        return 5
    }
}

// 커피 데코레이터 클래스 (기본 커피 객체를 감싸는 클래스)
open class CoffeeDecorator(protected val coffee: Coffee) : Coffee() {
    override fun cost(): Int {
        return coffee.cost()
    }
}

// 우유 데코레이터 클래스
class MilkDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost(): Int {
        return coffee.cost() + 2  // 우유는 2가 추가됨
    }
}

// 설탕 데코레이터 클래스
class SugarDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost(): Int {
        return coffee.cost() + 1  // 설탕은 1이 추가됨
    }
}

fun main() {
    val coffee = Coffee()
    println(coffee.cost())  // 5

    val coffeeWithMilk = MilkDecorator(coffee)
    println(coffeeWithMilk.cost())  // 7

    val coffeeWithSugar = SugarDecorator(coffee)
    println(coffeeWithSugar.cost())  // 6

    val coffeeWithMilkAndSugar = SugarDecorator(coffeeWithMilk)
    println(coffeeWithMilkAndSugar.cost())  // 8
}
